app/src/main/scala/org/alephium/explorer/persistence/model/BlockHeader.scala

```diff
@@ -65,4 +67,62 @@ object BlockHeader {
       blockEntity.hashrate,
       blockEntity.parent(groupNum)
     )
+
+  /**
+    * Finds a block with maximum height. If two blocks have the same
+    * height, returns the one with maximum timestamp.
+    * */
+  def maxHeight(blocks: Iterable[BlockHeader]): Option[BlockHeader] =
+    blocks.foldLeft(Option.empty[BlockHeader]) {
+      case (currentMax, header) =>
+        currentMax match {
+          case Some(current) =>
+            if (header.height > current.height) {
+              Some(header)
+            } else if (header.height == current.height) {
+              maxTimeStamp(Array(current, header))
+            } else {
+              currentMax
+            }
+
+          case None =>
+            Some(header)
+        }
+    }
+
+  /** Finds a block with maximum timestamp */
+  def maxTimeStamp(blocks: Iterable[BlockHeader]): Option[BlockHeader] =
+    blocks.foldLeft(Option.empty[BlockHeader]) {
+      case (currentMax, header) =>
+        currentMax match {
+          case Some(current) =>
+            if (header.timestamp > current.timestamp) {
+              Some(header)
+            } else {
+              currentMax
+            }
+
+          case None =>
+            Some(header)
+        }
+    }
+
+  /** Filter blocks within the given chain */
+  @inline def filterBlocksInChain(blocks: Iterable[BlockHeader],
+                                  chainFrom: GroupIndex,
+                                  chainTo: GroupIndex): Iterable[BlockHeader] =
+    blocks.filter { header =>
+      header.chainFrom == chainFrom &&
+      header.chainTo == chainTo
+    }
+
+  /** Sum height of all blocks */
+  def sumHeight(blocks: Iterable[BlockHeader]): Option[Height] =
+    blocks.foldLeft(Option.empty[Height]) {
+      case (currentSum, header) =>
+        currentSum match {
+          case Some(sum) => Some(Height.unsafe(sum.value + header.height.value))
+          case None      => Some(header.height)
+        }
+    }
```

Comment:
all above functions are only used in tests, I would rather move them there

Reply:
I wasn't sure where in test you'd preferred them to be. I've placed them in `BlockHeaderSpec`. Let me know if you'd like them moved elsewhere.